madsim-aws-sdk-s3/src/server/service.rs

```diff
@@ -0,0 +1,624 @@
+            } else if range_set.ends_with('-') {
+                let end_pos = range_set
+                    .split('-')
+                    .next()
+                    .ok_or_else(|| Error::InvalidRangeSpecifier(range.clone()))?
+                    .parse::<usize>()
+                    .map_err(|_| Error::InvalidRangeSpecifier(range.clone()))?;
+
+                body[..end_pos].to_vec()
+            } else {
+                let first_pos = range_set
+                    .split('-')
+                    .next()
+                    .ok_or_else(|| Error::InvalidRangeSpecifier(range.clone()))?
+                    .parse::<usize>()
+                    .map_err(|_| Error::InvalidRangeSpecifier(range.clone()))?;
+
+                let end_pos = range_set
+                    .split('-')
+                    .next()
+                    .ok_or_else(|| Error::InvalidRangeSpecifier(range.clone()))?
+                    .parse::<usize>()
+                    .map_err(|_| Error::InvalidRangeSpecifier(range.clone()))?;
+
+                body[first_pos..end_pos].to_vec()
+            };
+
+            let body = crate::types::ByteStream::from(body);
+            Ok(GetObjectOutput { body })
+        } else if let Some(part_number) = part_number {
+            let body = object.body.clone();
+
+            if part_number >= 0 {
+                let part_number = part_number as usize;
+                if part_number < body.len() {
+                    let body =
+                        crate::types::ByteStream::from(body[part_number..part_number + 1].to_vec());
+                    return Ok(GetObjectOutput { body });
+                }
+            }
+            Err(Error::InvalidPartNumberSpecifier(part_number))
+        } else {
+            Ok(GetObjectOutput {
+                body: crate::types::ByteStream::from(object.body.clone()),
+            })
+        }
+    }
+
+    async fn put_object(
+        &mut self,
+        bucket: String,
+        key: String,
+        body: crate::types::ByteStream,
+    ) -> Result<PutObjectOutput> {
+        let object = self
+            .storage
+            .get_mut(&bucket)
+            .ok_or(Error::InvalidBucket(bucket))?
+            .entry(key)
+            .or_default();
+
+        let body = body.collect().await;
+        let body = body.expect("error read data").into_bytes().to_vec();
+
+        object.body = body;
+        object.completed = true;
+
+        Ok(PutObjectOutput {})
+    }
+
+    fn delete_object(&mut self, bucket: String, key: String) -> Result<DeleteObjectOutput> {
+        let object = self
+            .storage
+            .get_mut(&bucket)
+            .ok_or(Error::InvalidBucket(bucket))?
+            .get_mut(&key)
+            .ok_or_else(|| Error::InvalidKey(key.clone()))?;
+
+        if !object.completed {
+            Err(Error::InvalidKey(key))
+        } else {
+            object.completed = false;
+            object.body.clear();
```

Comment:
Why not directly remove the key? Is that for the case when there's uploading at the same time?

Reply:
yes, put the multipart under the object is just for convenience, so should not clear the parts under the object, if there are some multipart upload requests